src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/ComService.kt

```diff
@@ -12,26 +13,64 @@ class ComService(
   private val communityOffenderManagerRepository: CommunityOffenderManagerRepository,
 ) {
 
+  companion object {
+    private val log = LoggerFactory.getLogger(this::class.java)
+  }
+
+  /**
+   * Check if record already exists. If so, check if any of the details have changed before performing an update.
+   * Check if the username and staffId do not match. This should not happen, unless a Delius account is updated to point
+   * at another linked account using the staffId. In this scenario, we should update the existing record to reflect
+   * the new username and or staffId.
+   */
   @Transactional
   fun updateComDetails(comDetails: UpdateComRequest): CommunityOffenderManager {
-    val com = this.communityOffenderManagerRepository.findByStaffIdentifier(comDetails.staffIdentifier)
-    val updatedCom = if (com !== null) {
-      com.copy(
-        email = comDetails.staffEmail,
-        firstName = comDetails.firstName,
-        lastName = comDetails.lastName,
-        lastUpdatedTimestamp = LocalDateTime.now()
+    val comResult = this.communityOffenderManagerRepository.findByStaffIdentifierOrUsername(
+      comDetails.staffIdentifier, comDetails.staffUsername
+    )
+
+    if (comResult == null || comResult.isEmpty()) {
+      return this.communityOffenderManagerRepository.saveAndFlush(
+        CommunityOffenderManager(
+          username = comDetails.staffUsername,
+          staffIdentifier = comDetails.staffIdentifier,
+          email = comDetails.staffEmail,
+          firstName = comDetails.firstName,
+          lastName = comDetails.lastName
+        )
       )
-    } else {
-      CommunityOffenderManager(
-        username = comDetails.staffUsername,
-        staffIdentifier = comDetails.staffIdentifier,
-        email = comDetails.staffEmail,
-        firstName = comDetails.firstName,
-        lastName = comDetails.lastName
+    }
+
+    if (comResult.count() > 1) {
+      log.warn(
+        "More then one COM record found for staffId {} username {}",
+        comDetails.staffIdentifier,
+        comDetails.staffUsername
+      )
```

Comment:
Good question. We can receive "event bombs" where we have many event fired within a second or so. In this instance, the API can throw a lot of exceptions in quick succession.

When the DB was throwing exceptions on the key constraint, the app become very slow and unresponsive. Since exceptions are heavy to process, I opted for a log in place of an exception. But it could be changed to just return with no updates. 

I need to monitor for this scenario, as it would require Delius to consolidate the two accounts into one. Unless theres a legitimate reason for having more than one account with the same username and or staffId